### PR TITLE
UniqThetaSketch support set operation such as union/intersect/not

### PR DESCRIPTION
**Pull request by pzhdfy**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
after https://github.com/ClickHouse/ClickHouse/issues/14893 
clickhouse has datasketches aggregation support.
but set operation such as ∩ ∪ × is not implement.

In our scene,we want to do Retention Analysis. but retention function in clickhouse is accurate and costs a lot.
we need an approximate method instead.

Druid has a implement using datasketches
https://druid.io/docs/0.12.3/development/extensions-core/datasketches-aggregators.html
The key implementition is **support  intersect operation between datasketches objects**

other issues:https://github.com/ClickHouse/ClickHouse/issues/38842

so we add **_uniqThetaIntersect_** function ( bonus  **_uniqThetaUnion_** ,**_uniqThetaNot_**)

typic usage:
to compute  (uv2022 ∩ uv2021)/uv2021
```sql
CREATE TABLE test2
(
    `year` String ,
    `uv`  Int64
)
ENGINE = MergeTree()
ORDER BY (year);

INSERT INTO TABLE test2(year, uv) select '2021',1;
INSERT INTO TABLE test2(year, uv) select '2021',2;
INSERT INTO TABLE test2(year, uv) select '2021',3;
INSERT INTO TABLE test2(year, uv) select '2021',4;
INSERT INTO TABLE test2(year, uv) select '2022',1;
INSERT INTO TABLE test2(year, uv) select '2022',3;

select finalizeAggregation(uniqThetaIntersect(uv2021,uv2022))/finalizeAggregation(uv2021)
from
(
select uniqThetaStateIf(uv,year='2021') as uv2021, uniqThetaStateIf(uv,year='2022') as uv2022 
from test2
);
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by UnamedRus** [comment]
UNION actually exist:

```
WITH
    uniqThetaState(number) AS a,
    uniqThetaState(number + 10) AS b
SELECT finalizeAggregation(a + b)
FROM numbers(10) AS result

┌─finalizeAggregation(plus(a, b))─┐
│                              20 │
└─────────────────────────────────┘
```

File: src/Functions/UniqTheta/FunctionsUniqTheta.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+            const auto * sketch_type1 = typeid_cast<const DataTypeAggregateFunction *>(arguments[1].get());
+            if (!(sketch_type1 && sketch_type1->getFunctionName() == "uniqTheta"))
+                throw Exception(
+                        "Second argument for function " + getName() + " must be a uniqTheta but it has type " + arguments[1]->getName(),
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+            //todo:check
+            if (sketch_type0->getArgumentsDataTypes()[0]->getTypeId() != sketch_type1->getArgumentsDataTypes()[0]->getTypeId())
```
Comment (Avogar) [inline src/Functions/UniqTheta/FunctionsUniqTheta.h:90]: `DataType` class has method `equals`. And also, why do we check only the first argument?

File: src/AggregateFunctions/ThetaSketchData.h [hunk 2]
```diff
@@ -80,6 +82,58 @@ class ThetaSketchData : private boost::noncopyable
...
+            theta_intersection.update(*rhs.sk_update);
+        else if (rhs.sk_union)
+            theta_intersection.update(rhs.sk_union->get_result());
+
+        sk_union.reset(nullptr);
+        u = getSkUnion();
+        u->update(theta_intersection.get_result());
+    }
+
+    void a_not_b(const ThetaSketchData & rhs)
```
Comment (Avogar) [inline src/AggregateFunctions/ThetaSketchData.h:109]: Let's use `camelCase` style for methods naming

**Review by Avogar — APPROVED** [review]
Just two small comments

**Comment by pzhdfy** [comment]
> UNION actually exist:
> 
> ```
> WITH
>     uniqThetaState(number) AS a,
>     uniqThetaState(number + 10) AS b
> SELECT finalizeAggregation(a + b)
> FROM numbers(10) AS result
> 
> ┌─finalizeAggregation(plus(a, b))─┐
> │                              20 │
> └─────────────────────────────────┘
> ```

i found another trick  style(🙂):
```
WITH
    uniqThetaState(number) AS a,
    uniqThetaState(number + 10) AS b
SELECT finalizeAggregation(arrayReduce('uniqThetaMergeState',[a,b]))
FROM numbers(10) AS result;
```

File: src/Functions/UniqTheta/FunctionsUniqTheta.h (hunk 1 above)
Comment (Avogar) [reply src/Functions/UniqTheta/FunctionsUniqTheta.h:90]: @pzhdfy

Comment (pzhdfy) [reply src/Functions/UniqTheta/FunctionsUniqTheta.h]: update.   check all ArgumentsDataTypes and use equals

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
@pzhdfy Please, look at the AST fuzzer failure, it's related to the changes

**Comment by pzhdfy** [comment]
> @pzhdfy Please, look at the AST fuzzer failure, it's related to the changes

I can not reproduce with this sql.

2022.08.23 12:40:30.272410 [ 131 ] {} <Trace> BaseDaemon: Received signal -3
2022.08.23 12:40:30.272804 [ 448 ] {} <Fatal> BaseDaemon: ########################################
2022.08.23 12:40:30.272895 [ 448 ] {} <Fatal> BaseDaemon: (version 22.9.1.1, build id: 7AB196498DF66E0B) (from thread 206) (query_id: a26452f5-4f0b-4284-b129-42f5de4b81b7) (query: SELECT finalizeAggregation(uniqThetaIntersect(a, b)), uniqThetaNot(a, b), finalizeAggregation(a), NULL FROM (SELECT arrayReduce(arrayReduce(NULL, [7, 1048576, 1, -2147483648], [NULL, NULL, -2147483649, NULL]), NULL, [NULL, NULL]), uniqThetaStateIf(number, number > 1023) AS a, uniqThetaStateIf(number, number > -2147483649) AS b FROM (SELECT arrayReduce(arrayReduce('uniqThetaState', [1023, -1]), arrayReduce(NULL, [NULL, NULL, -2147483649]), NULL, []), number FROM system.numbers LIMIT 1048575))) Received signal Unknown signal (-3)
2022.08.23 12:40:30.272933 [ 448 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2022.08.23 12:40:30.272989 [ 448 ] {} <Fatal> BaseDaemon: Stack trace: 0x10292e23 0x10528676 0x10249286 0x1025c4b8 0x148f9161 0x148f884a 0x24aa975b 0x26fd7b86 0x26fd73f3 0x26f946d5 0x26f9082a 0x26f821f4 0x26fa22ba 0x27ee7e0c 0x27ee82fa 0x2805ffe7 0x2805dacf 0x7fb8e0647609 0x7fb8e056c133

and I review the error statck, it happens in  DB::ThetaSketchData<unsigned long>::write,  may be it is a rare bug in datasketches.
I add a blank line , to rerun the check,  and AST fuzzer runs well this time.

2022.08.23 12:40:30.097297 [ 278 ] {} <Trace> SystemLog (system.metric_log): Flushed system log up to offset 189
    #0 0x148f9160 in datasketches::copy_to_mem(void const*, void*, unsigned long) build_docker/../contrib/datasketches-cpp/common/include/memory_operations.hpp:51:3
    #1 0x148f9160 in datasketches::compact_theta_sketch_alloc<std::__1::allocator<unsigned long> >::serialize(unsigned int) const build_docker/../contrib/datasketches-cpp/theta/include/theta_sketch_impl.hpp:391:12
    #2 0x148f8849 in DB::ThetaSketchData<unsigned long>::write(DB::WriteBuffer&) const build_docker/../src/AggregateFunctions/ThetaSketchData.h:158:49
    #3 0x24aa975a in DB::SerializationAggregateFunction::serializeBinaryBulk(DB::IColumn const&, DB::WriteBuffer&, unsigned long, unsigned long) const build_docker/../src/DataTypes/Serializations/SerializationAggregateFunction.cpp:73:19
    #4 0x26fd7b85 in DB::writeData(DB::ISerialization const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn> const&, DB::WriteBuffer&, unsigned long, unsigned long) build_docker/../src/Formats/NativeWriter.cpp:63:19
    #5 0x26fd73f2 in DB::NativeWriter::write(DB::Block const&) build_docker/../src/Formats/NativeWriter.cpp:164:13
    #6 0x26f946d4 in DB::TCPHandler::sendData(DB::Block const&) build_docker/../src/Server/TCPHandler.cpp:1742:26
    #7 0x26f90829 in DB::TCPHandler::processOrdinaryQueryWithProcessors() build_docker/../src/Server/TCPHandler.cpp:748:21
    #8 0x26f821f3 in DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:368:17
    #9 0x26fa22b9 in DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:1866:9
    #10 0x27ee7e0b in Poco::Net::TCPServerConnection::start() build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3
    #11 0x27ee82f9 in Poco::Net::TCPServerDispatcher::run() build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #12 0x2805ffe6 in Poco::PooledThread::run() build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14
    #13 0x2805dace in Poco::ThreadImpl::runnableEntry(void*) build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27
    #14 0x7fb8e0647608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477:8
    #15 0x7fb8e056c132 in __clone /build/glibc-SzIz7B/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95

**Comment by Avogar** [comment]
> I can not reproduce with this sql.

It reproduces easily, you just need to use binary that was built with ubsan (you can download it from 
ClickHouse build check report).

> I add a blank line , to rerun the check, and AST fuzzer runs well this time.

It's because ASTFuzzer uses randomization, this time such query that causes this error just wasn't generated.

> and I review the error statck, it happens in DB::ThetaSketchData::write, may be it is a rare bug in datasketches.

Even if the problem is in datasketches library, we cannot merge this PR until this bug is fixed. We use our own fork of datasketches library (https://github.com/ClickHouse/datasketches-cpp/tree/master), so you can try to find and fix this bug and create a PR (also maybe it was already fixed in original datasketches repository, need to check).

**Comment by pzhdfy** [comment]
I will try to fix it

**Comment by pzhdfy** [comment]
> I made reproduce code with dataskethes. and submit an issue: [apache/datasketches-cpp#300](https://github.com/apache/datasketches-cpp/issues/300)

fixed by upstream, and I made a PR to merge to our forked datasketches-cpp.

https://github.com/ClickHouse/datasketches-cpp/pull/3

**Comment by Avogar** [comment]
> fixed by upstream, and I made a PR to merge to our forked datasketches-cpp.

Great, now let's update reference commit in contrib directory in this PR and if tests will be OK we can merge it

File: src/Functions/UniqTheta/FunctionsUniqTheta.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#if USE_DATASKETCHES
+
+namespace DB
+{
+
+REGISTER_FUNCTION(UniqTheta)
+{
+    factory.registerFunction<FunctionUniqThetaIntersect>();
+    factory.registerFunction<FunctionUniqThetaUnion>();
+    factory.registerFunction<FunctionUniqThetaNot>();
```
Comment (Avogar) [inline src/Functions/UniqTheta/FunctionsUniqTheta.cpp:14]: After recent changes in https://github.com/ClickHouse/ClickHouse/pull/40711 all new functions must have an embedded documentation (see new argument `doc` in registerFunction method). Please, add documentation for these functions

**Pull request merged**